File: Ps2/extract_products.js
const XLSX = require('xlsx');
const path = require('path');

try {
    const filePath = path.join(__dirname, 'STOCK.xlsx');
    const workbook = XLSX.readFile(filePath);
    const worksheet = workbook.Sheets[workbook.SheetNames[0]];
    const data = XLSX.utils.sheet_to_json(worksheet);
    
    const products = [];
    const grouped = {};
    
    data.forEach(row => {
        const name = row['Product Name'] || row['product_name'] || Object.values(row)[0];
        const category = row['Category'] || row['category'] || Object.values(row)[1];
        let price = row['Price'] || row['price'] || Object.values(row)[2];
        let quantity = row['Quantity'] || row['quantity'] || Object.values(row)[3];
        
        if (!name) return;
        
        if (typeof price === 'string') {
            price = parseFloat(price.replace('₹', '').trim());
            if (isNaN(price)) price = 0;
        }
        
        if (quantity === '' || quantity === null || quantity === undefined) {
            quantity = 0;
        } else {
            quantity = parseInt(quantity);
            if (isNaN(quantity)) quantity = 0;
        }
        
        const productDict = {
            name: String(name).trim(),
            category: String(category).trim(),
            price: price,
            quantity: quantity
        };
        
        products.push(productDict);
        
        const categoryKey = String(category).trim();
        if (!grouped[categoryKey]) {
            grouped[categoryKey] = [];
        }
        grouped[categoryKey].push(productDict);
    });
    
    const result = {
        products: products,
        grouped: grouped,
        total_products: products.length,
        total_categories: Object.keys(grouped).length
    };
    
    console.log(JSON.stringify(result, null, 2));
} catch (error) {
    console.log(JSON.stringify({ error: error.message }, null, 2));
}

File: Ps2/extract_products.py
import json
from pathlib import Path

try:
    import openpyxl
    wb = openpyxl.load_workbook(r'C:\Users\pagar\Desktop\Ps2\STOCK.xlsx')
    ws = wb.active
    
    rows = list(ws.iter_rows(values_only=True))
    headers = rows[0]
    
    products = []
    grouped = {}
    
    for row in rows[1:]:
        if not any(row):
            continue
            
        product = {}
        for i, header in enumerate(headers):
            product[header.lower().replace(' ', '_')] = row[i]
        
        name = product.get('product_name', '')
        category = product.get('category', '')
        price = product.get('price', '')
        quantity = product.get('quantity', '')
        
        if name:
            if isinstance(price, str):
                price = price.replace('₹', '').strip()
                try:
                    price = float(price)
                except:
                    price = price
            
            if quantity == '' or quantity is None:
                quantity = 0
            else:
                try:
                    quantity = int(quantity)
                except:
                    quantity = 0
            
            product_dict = {
                "name": name,
                "category": category,
                "price": price,
                "quantity": quantity
            }
            products.append(product_dict)
            
            if category not in grouped:
                grouped[category] = []
            grouped[category].append(product_dict)
    
    result = {
        "products": products,
        "grouped": grouped,
        "total_products": len(products),
        "total_categories": len(grouped)
    }
    
    print(json.dumps(result, ensure_ascii=False, indent=2))

except Exception as e:
    try:
        import pandas as pd
        df = pd.read_excel(r'C:\Users\pagar\Desktop\Ps2\STOCK.xlsx')
        
        products = []
        grouped = {}
        
        for _, row in df.iterrows():
            name = row.iloc[0]
            category = row.iloc[1]
            price = row.iloc[2]
            quantity = row.iloc[3]
            
            if pd.isna(name):
                continue
            
            if isinstance(price, str):
                price = price.replace('₹', '').strip()
                try:
                    price = float(price)
                except:
                    price = price
            
            if pd.isna(quantity) or quantity == '':
                quantity = 0
            else:
                try:
                    quantity = int(quantity)
                except:
                    quantity = 0
            
            product_dict = {
                "name": str(name),
                "category": str(category),
                "price": price,
                "quantity": quantity
            }
            products.append(product_dict)
            
            category_key = str(category)
            if category_key not in grouped:
                grouped[category_key] = []
            grouped[category_key].append(product_dict)
        
        result = {
            "products": products,
            "grouped": grouped,
            "total_products": len(products),
            "total_categories": len(grouped)
        }
        
        print(json.dumps(result, ensure_ascii=False, indent=2))
    except Exception as e2:
        print(json.dumps({"error": str(e2)}, ensure_ascii=False))
